doc/design/codegen_couler_use_ir.md

```diff
@@ -0,0 +1,60 @@
+# Use IR to Generate Couler Program
+
+## Background
+
+We are planning to migrate SQLFlow submitters to use Couler: https://github.com/sql-machine-learning/sqlflow/pull/1146/.
+
+In this design, we'll only have `codegen_couler.go` to generate programs to run a SQL program, which will be translated into multiple couler "steps". Each step, we can execute a SQL statement, run a training python program or run prediction, etc.
+
+## Generate Couler Program Using IR
+
+For example, to run the training steps, we intend to use couler API call: `couler.{xgboost,tensorflow,elasticdl}.train(model_def, data)` to generate a training step. The `train` function must take lots of parameters to run the training job as the SQLFlow statement describes, see [here](https://github.com/sql-machine-learning/sqlflow/blob/develop/python/sqlflow_submitter/tensorflow/train.py#L52) as an example.
+
+To implement the single `codegen_couler.go` to support generate code that can run either Tensorflow/XGBoost/ElasticDL/ALPS programs, we have below two choices:
+
+1. `couler.{tensorflow/xgboost/elasticdl}.train` have different arguments defined, so we can do:
+
+    ```go
+        if ir.ModelType == "Tensorflow":
+            tfFiller := TFFiller{
+                Estimator: generateTFEstimatorCode(ir),
+                FeatureColumns: generateFeatureColumnsCode(ir),
+                FieldMetas: generateFieldMetasCode(ir),
+                ...
+            }
+            tfTrainTemplate = `couler.tensorflow.train(estimator="{{.Estimator}}", FeatureColumns="""{{.FeatureColumns}}""", FieldMetas={{.FieldMetas}})`
+            // Do template rendering here.
+        else if ir.ModelType == "XGBoost":
+            ...
+    ```
+2. `couler.{tensorflow/xgboost/elasticdl}.train` accept a parameter called `train_config` which is a python dictionary that can be generated by SQLFlow IR.
+
+    ```go
+        filler := Filler{
+            ModelType: ir.ModelType,
+            TrainConfigCode : generateTrainConfig(ir),
+        }
+        tfTrainTemplate = `couler.{{.ModelType}}.train(train_config={{.TrainConfigCode}})`
+        // Do template rendering here.
+    ```
+    **Note** that the `train_config` dictionary is not "engine specific", so that it could be used by any engine. A sample `train_config` should look like:
```

Comment:
LGTM. In this design, SQLFlow is only responsible for the parsing, verifying, feature derivation, and serializing the result into the following Python struct.

Reply:
And we need to think about who is responsible for generating the user documentation. In my opinion, who is translating `train.epoch` to `epoch` should be responsible for docgen, which is Couler in this design. Do you think it is reasonable?

Reply:
@tonyyang-svail can't agree more. By the way, if couler is generating the documentation, we should put couler repo under github.com/sql-machine-learning so that we can generate documentation under sqlflow.org.